### PR TITLE
feat(cli): auto-generate index.ts for exports

### LINKED ISSUE #1127: [CLI] Generating `index.ts` in each models/repositories/controllers automatically
## Description / Steps to reproduce / Feature proposal
In an LB4 app (e.g. in `example-getting-started` example), it seems like the general pattern is to have the export statements in the `index.ts` file in each folders (models/repositories/controllers), so that the classes in those folders can simply be imported by 
`import {SomeClass} from "../models";`

## Acceptance Criteria
I wonder if this is a good idea to:
- generate those index.ts automatically as part of `lb4 app` command
- for any artifacts created by CLI (currently we only have `lb4 controller`), add the export in index.ts
e.g. `export * from './someclass.repository';`

### PR DESCRIPTION
**Pull request by virkt25**
This is part of the work I did for the **Boot Spike**. `lb4 datasource` makes use of the index generation so decided to just deliver this as it's own feature. 

fix #1127

**UPDATE: MAY 31**
- The PR no longer re-generated index.ts -- If already present, it adds a line at the end. If it doesn't exist in the folder it creates it and adds the file to to it.
- Auto-generation can be skipped as documented in code

## Checklist

- [x] `npm test` passes on your machine
- [ ] New tests added or existing tests modified to cover all changes
- [ ] Code conforms with the [style guide](http://loopback.io/doc/en/contrib/style-guide.html)
- [ ] API Documentation in code was updated
- [x] Documentation in [/docs/site](../tree/master/docs/site) was updated
- [x] Affected artifact templates in `packages/cli` were updated
- [x] Affected example projects in `examples/*` were updated

### CODE REVIEW

File: packages/cli/generators/controller/index.js [hunk 1]
```diff
@@ -235,6 +235,7 @@ module.exports = class ControllerGenerator extends ArtifactGenerator {
         'Controller %s is now created in src/controllers/',
         this.artifactInfo.name,
       );
+      super.updateIndexFile();
```
Comment (shimks) [inline packages/cli/generators/controller/index.js]: Should this be returned since it's a promise? It might not matter since it's the last thing being done here, but I think it makes sense to return it for the sake of using a promise

File: packages/cli/lib/make-index.js [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+  const code = format(lines.join('\n'));
+
+  debug('===== code =====');
+  debug(code);
+  debug('===== code =====');
+
+  const outFile = `${dir}/index.ts`;
+  debug(`outFile => ${outFile}`);
+
+  const writeFileAsync = util.promisify(fs.writeFile);
```
Comment (shimks) [inline packages/cli/lib/make-index.js]: Is there any reason as to why we don't use `fs.writeFileSync`?

File: packages/cli/lib/artifact-generator.js [hunk 3]
```diff
@@ -102,4 +104,16 @@ module.exports = class ArtifactGenerator extends BaseGenerator {
       {globOptions: {dot: true}},
     );
   }
+
+  async updateIndexFile() {
+    await index(this.artifactInfo.outDir, {
+      prefix: `.${this.artifactInfo.type}`,
```
Comment (shimks) [inline packages/cli/lib/artifact-generator.js]: It's weird to me to have to `.` as part of `prefix`. I think it's better for `.` to be inferred when setting `globPattern` in `make-index.js`

File: packages/cli/generators/controller/index.js (hunk 1 above)
Comment (shimks) [reply packages/cli/generators/controller/index.js]: Also, shouldn't `updateIndexFile` be called in `ArtifactGenerator` since most artifacts seem to use `index.ts`?

File: examples/todo/src/controllers/index.ts [hunk 4]
```diff
@@ -1,6 +1,5 @@
-// Copyright IBM Corp. 2017,2018. All Rights Reserved.
-// Node module: @loopback/example-todo
-// This file is licensed under the MIT License.
-// License text available at https://opensource.org/licenses/MIT
```
Comment (bajtos) [inline examples/todo/src/controllers/index.ts]: Shouldn't the example files preserve license headers? (I understand the files generated by `lb4` do not come with license headers.)

File: packages/cli/.prettierrc [hunk 5]
```diff
@@ -1,6 +0,0 @@
-{
-  "bracketSpacing": false,
```
Comment (bajtos) [inline packages/cli/.prettierrc]: What is the reason for removing `.prettierrc` from this project? I am not against such change per se, but would like to better understand the reasoning. It may be better to make this change in a standalone pull request.

File: packages/cli/lib/format.js [hunk 6]
```diff
@@ -0,0 +1,13 @@
+// Copyright IBM Corp. 2017,2018. All Rights Reserved.
+// Node module: @loopback/cli
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+const prettier = require('prettier');
```
Comment (bajtos) [inline packages/cli/lib/format.js]: _Cross-posting item 8 from https://github.com/strongloop/loopback-next/issues/441#issuecomment-392522724:_

Formatting of generated code: it would be great if our tooling was able to pick up prettier version and configuration from the project we are running in and format the generated code accordingly.

Imagine a user is using `prettier` in their project (as scaffolded by `lb4 app`), but changed prettier config to use double-quotes and tabs. If we don't honor their style, they will have to run `npm run lint:fix` every time they add a new model or update auto-generated TypeScript definitions.

It is also important to run `prettier` from `node_modules` of the target project, because different prettier version have variations in how they format the same piece of code and we cannot get all users to have exactly the same version of prettier as we use in LB4 tooling.

It may be enough to add `${this.destinationPath()}/node_modules` to the list of directories where Node.js looks for modules. See how Mocha handles that ([link](https://github.com/mochajs/mocha/blob/0d95e3f41a0859be4e7225541575e78b6d8926ee/bin/_mocha#L327)):

```ts
module.paths.push(cwd, join(cwd, 'node_modules'));
```

I am ok to leave these two improvements out of scope of this initial pull request as long as there is a follow-up issue (or two?) to ensure we fix this problem as part of the DP3 Boot epic.

File: packages/cli/lib/make-index.js [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+  // Only exporting typescript files.
+  // TODO: @virkt25 -- May want this to be configurable in the future?
+  const EXTENSION = '.ts';
+
+  // Default Options for Making an exports index file
+  const defaultOptions = {
+    nested: true,
+    prefix: '',
+    commentHeader: `This is an auto-generated file. DO NOT EDIT.`,
+    commentFooter: `This is an auto-generated file. DO NOT EDIT.`,
```
Comment (bajtos) [inline packages/cli/lib/make-index.js]: Is there any particular reason for using backticks (template literals) instead of regular single-quotes?

File: packages/cli/lib/make-index.js [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+  // Add comment header as first line
+  if (options.commentHeader) {
+    lines.push(normalizeComment(options.commentHeader));
+  }
+
+  // For each file discovers by the glob pattern, generate an export statement
+  files.forEach(file => {
+    const relPath = `export * from './${path
+      .relative(dir, file)
+      .slice(0, -3)}';`;
```
Comment (bajtos) [inline packages/cli/lib/make-index.js]: Please extract the expression computing the actual path outside of the template literal and save the result to a variable.

```ts
const relativePath = path.relative(dir, file).slice(0, -3);
const exportLine = `export * from '${relativePath}';`;
lines.push(relPath);
```

File: packages/cli/lib/make-index.js [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+
+  debug('===== code =====');
+  debug(code);
+  debug('===== code =====');
+
+  const outFile = `${dir}/index.ts`;
+  debug(`outFile => ${outFile}`);
+
+  const writeFileAsync = util.promisify(fs.writeFile);
+  await writeFileAsync(outFile, code);
```
Comment (bajtos) [inline packages/cli/lib/make-index.js]: This will break incremental builds, because every run of `make-index` is going to change the modification time of `index.ts` to the current date, even in cases where the file content was not changed at all.

Proposed solution:

 1. Read the current file content and if the old content is the same as the new content, then don't write anything. I believe this is what prettier is doing too.
 2. Set `mtime` of the written file to match the latest `mtime` of files we were generating index from. I.e. for each file in `dir` (excluding `index.ts`), obtain `mtime` value and then pick the maximum value of all mtimes.

I think the second item is not enough on its own, because file watchers like `tsc --watch`, fs-events, etc. listen for write events and ignore `mtime`.

The first item may is most likely good enough, because typically the `index.ts` file would be updated shortly after the last non-index file was added/updated.

File: packages/cli/lib/make-index.js [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+/**
+ * Takes a string and turns it into a comment -- existing comments are normalized
+ *
+ * @param {string} comment String to be converted to comment
+ */
+function normalizeComment(comment) {
+  const comments = comment.split('\n');
+  let normalizedComment = '';
+
+  // For each comment line -- if it's a comment we
```
Comment (bajtos) [inline packages/cli/lib/make-index.js]: > if it's a comment we

We do what? Looks like you accidentally forgot to finish the sentence here.

File: packages/cli/lib/make-index.js (hunk 9 above)
Comment (bajtos) [reply packages/cli/lib/make-index.js]: I am ok to leave these two improvements out of scope of this initial pull request as long as there is a follow-up issue to ensure we fix this problem as part of the DP3 Boot epic.

File: packages/cli/.prettierrc (hunk 5 above)
Comment (virkt25) [reply packages/cli/.prettierrc]: The reason was that this `.prettierrc` uses different formatting that the top-level one we use for the project and I noticed that and removed it. I've reverted this change and will make a new PR for this.

File: packages/cli/generators/app/templates/src/controllers/index.ts.ejs [hunk 11]
```diff
@@ -0,0 +1 @@
+export * from './ping.controller';
```
Comment (shimks) [inline packages/cli/generators/app/templates/src/controllers/index.ts.ejs]: This makes me wonder whether it should be included when the app is generated as opposed to adding it on when the controller generator is run. Thoughts?

File: packages/cli/lib/update-index.js [hunk 12]
```diff
@@ -0,0 +1,23 @@
+// Copyright IBM Corp. 2017,2018. All Rights Reserved.
+// Node module: @loopback/cli
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+const path = require('path');
+const util = require('util');
+const fs = require('fs');
+const appendFileAsync = util.promisify(fs.appendFile);
```
Comment (shimks) [inline packages/cli/lib/update-index.js]: Why not use this instead? https://nodejs.org/api/fs.html#fs_fs_appendfilesync_path_data_options

File: packages/cli/lib/update-index.js [hunk 13]
```diff
@@ -0,0 +1,23 @@
...
+ * @param {String} dir The directory in which index.ts is to be updated/created
+ * @param {*} file The new file to be exported from index.ts
+ */
+module.exports = async function(dir, file) {
+  const indexFile = path.join(dir, 'index.ts');
+  if (!file.endsWith('.ts')) {
+    throw new Error(`${file} must be a TypeScript (.ts) file`);
+  }
+  const content = `export * from './${file.slice(0, -3)}';\n`;
+  await appendFileAsync(indexFile, content);
```
Comment (shimks) [inline packages/cli/lib/update-index.js]: Should we explicitly check whether the line exists already? Or is it that too much of an edge case

File: packages/cli/test/unit/update-index.unit.js [hunk 14]
```diff
@@ -0,0 +1,47 @@
...
+const expect = testlab.expect;
+const TestSandbox = testlab.TestSandbox;
+
+// Test Sandbox
+const SANDBOX_PATH = path.resolve(__dirname, '.sandbox');
+const sandbox = new TestSandbox(SANDBOX_PATH);
+const expectedFile = path.join(SANDBOX_PATH, 'index.ts');
+
+describe('update-index unit tests', () => {
+  beforeEach('reset sandbox', () => sandbox.reset());
```
Comment (shimks) [inline packages/cli/test/unit/update-index.unit.js]: Is it possible to just provide `sandbox.reset` since it isn't async anymore?

File: packages/cli/test/unit/update-index.unit.js [hunk 15]
```diff
@@ -0,0 +1,47 @@
...
+      `export * from './first';\n`,
+    );
+    await updateIndex(SANDBOX_PATH, 'test.ts');
+    assert.file(expectedFile);
+    assert.fileContent(expectedFile, /export \* from '.\/first'/);
+    assert.fileContent(expectedFile, /export \* from '.\/test'/);
+  });
+
+  it('throws an error when given a non-ts file', async () => {
+    expect(updateIndex(SANDBOX_PATH, 'test.js')).to.be.rejectedWith(
```
Comment (shimks) [inline packages/cli/test/unit/update-index.unit.js]: needs to be wrapped in an anonymous function

File: packages/cli/generators/app/templates/src/controllers/index.ts.ejs (hunk 11 above)
Comment (virkt25) [reply packages/cli/generators/app/templates/src/controllers/index.ts.ejs]: Check the file -- It's in the `app` generator and this is just the default `index.ts` that gets generated since we are generating `ping.controller` in new apps.

File: packages/cli/lib/update-index.js (hunk 12 above)
Comment (virkt25) [reply packages/cli/lib/update-index.js]: `appendFileSync` is a blocking / synchronous operation. `async / await` is a promise under the hood even though it appears to be synchronous when written.

From https://hackernoon.com/6-reasons-why-javascripts-async-await-blows-promises-away-tutorial-c7ec10518dd9
> Async/await makes asynchronous code look and behave a little more like synchronous code. This is where all its power lies.

In a generator you can call `updateIndex` without an `await` ... and have more code below -- and that additional code and continue running without being blocked essentially. 

Here's a Node.js article on blocking / non-blocking. https://nodejs.org/en/docs/guides/blocking-vs-non-blocking/

File: packages/cli/lib/update-index.js (hunk 13 above)
Comment (virkt25) [reply packages/cli/lib/update-index.js]: Maybe. My assumption is to let the generator handle that -- within a generator this function is triggered only when the `generationStatus` is not an error / `identical` file. So if the file exists this won't be called. -- That said another check doesn't hurt. 🤷‍♂️

File: packages/cli/test/unit/update-index.unit.js (hunk 14 above)
Comment (virkt25) [reply packages/cli/test/unit/update-index.unit.js]: I've tried and I get the following error message if I do `beforeEach(sandbox.reset)`:

```
1) update-index unit tests
       "before each" hook: reset for "creates index.ts when not present":
     TypeError: this.validateInst is not a function
      at Context.reset (packages/testlab/src/test-sandbox.ts:31:27)
```

File: packages/cli/test/unit/update-index.unit.js (hunk 15 above)
Comment (virkt25) [reply packages/cli/test/unit/update-index.unit.js]: Why?

File: packages/cli/test/unit/update-index.unit.js (hunk 14 above)
Comment (shimks) [reply packages/cli/test/unit/update-index.unit.js]: Looks like it has to do with scoping :(.
On the other hand, do you think `async () => await sandbox.reset()` is cleaner than `() => sandbox.reset())`?

File: packages/cli/generators/app/templates/src/controllers/index.ts.ejs (hunk 11 above)
Comment (shimks) [reply packages/cli/generators/app/templates/src/controllers/index.ts.ejs]: The ping controller I know about, but I was wondering about whether we should include this `index.ts` file when the app generator is run

File: packages/cli/lib/update-index.js (hunk 13 above)
Comment (shimks) [reply packages/cli/lib/update-index.js]: Nvm, I forgot about yeoman's own conflicter system

File: packages/cli/test/unit/update-index.unit.js (hunk 14 above)
Comment (virkt25) [reply packages/cli/test/unit/update-index.unit.js]: I like the current format -- it's how we use `sandbox.reset` in the beforeEach hooks of other tests. Consistency is good IMO
